metadataserver/inject/inject_linux.go

```diff
@@ -0,0 +1,309 @@
+	}
+}
+
+/*
+1. It saves the current namespace
+2. It opens the container's network namespace at TITUS_PID_1_DIR/ns/net
+3. It calls unshare, and creates an entirely new network namespace (intermediate network namespace).
+4. It creates a veth pair in the intermediate
+5. It adds the IP address 169.254.169.254 to the veth pair
+6. It binds to port 80 in the new netns created
+7. It moves a side of the veth into the container's network namespace
+8. It sets up a route in the container's netns
+9. It changes to the PID ns of the container
+10. It calls the metadata proxy code
+*/
+func Inject(ctx context.Context, pid1dir string, subsequentExe []string) error {
+	ctx, cancel := context.WithTimeout(ctx, maxTime)
+	defer cancel()
+	listener, err := setupNamespaces(ctx, pid1dir)
+	if err != nil {
+		return fmt.Errorf("Could not get listener: %w", err)
+	}
+
+	/*
+		file, err := listener.File()
+		if err != nil {
+			return fmt.Errorf("Could not get listener file: %w", err)
+		}
+
+		err = unix.Dup2(int(file.Fd()), 169)
+		if err != nil {
+			return fmt.Errorf("Could not clone FD %d into 169: %w", file.Fd(), err)
+		}*/
+
+	err = unix.Dup2(*listener, 169)
+	if err != nil {
+		return fmt.Errorf("Could not clone FD %d into 169: %w", *listener, err)
+	}
+
+	err = unix.Exec(subsequentExe[0], subsequentExe, os.Environ())
+	if err != nil {
+		panic(fmt.Errorf("Unable to execute %v: %w", subsequentExe, err))
+	}
+
+	panic("Something went wrong")
+}
+
+// This is a separate function in order to allow for the defer / close statements to execute prior to calling exec
+// upon error, the failure state is undefined. Otherwise, on return, it should return to the host namespace.
+func setupNamespaces(ctx context.Context, pid1dir string) (*int, error) { // nolint: gocyclo
+	hostns, err := netns.Get()
+	if err != nil {
+		return nil, fmt.Errorf("Could not get current net ns: %w", err)
+	}
+	defer hostns.Close()
+	hostnsHandle, err := netlink.NewHandleAt(hostns)
+	if err != nil {
+		return nil, fmt.Errorf("Could not get handle at new host ns: %w", err)
+	}
+	defer hostnsHandle.Delete()
+
+	intermediateNS, err := netns.New()
+	if err != nil {
+		return nil, fmt.Errorf("Could not create new intermediate namespace: %w", err)
+	}
+	defer intermediateNS.Close()
+	intermediateNSHandle, err := netlink.NewHandleAt(intermediateNS)
+	if err != nil {
+		return nil, fmt.Errorf("Could not get handle at intermediate ns: %w", err)
+	}
+	defer intermediateNSHandle.Delete()
+
+	path := filepath.Join(pid1dir, "ns", "net")
+	containerNSFD, err := unix.Open(path, os.O_RDONLY, 0)
+	if err != nil {
+		return nil, fmt.Errorf("Could not open %q: %w", path, err)
+	}
+	defer unix.Close(containerNSFD)
+	containerNS := netns.NsHandle(containerNSFD)
+
+	containerNSHandle, err := netlink.NewHandleAt(containerNS)
+	if err != nil {
+		return nil, fmt.Errorf("Could not get handle at intermediate ns: %w", err)
+	}
+	defer containerNSHandle.Delete()
+
+	err = intermediateNSHandle.LinkAdd(&netlink.Veth{
```

Comment:
Previously, in the C code, adding veth links (probably not this line of code) was done in a for loop 5 times, do you have more faith that in golang it will either work or not here?

(I may be commenting on the wrong line of code)

Reply:
The loop was to workaround a kernel race, which I haven't had time to track down. :(  So I'm guessing it could still happen.